### ISSUE TITLE
Pktgen-DPDK  send packets failed

### ISSUE THREAD

**Issue by hufangyin** [pktgen/Pktgen-DPDK#44]
Hi
I use the Pktgen-DPDK in the VM, the Pktgen version is 2.9.5 and DPDK version is 2.1.0.  I have a Ruby Rapid Card in my host and use the PCI-passthrough to assign a PF to the VM. Here is my output:
[root@RRC1 sh]# lspci | grep Eth
04:00.0 Ethernet controller: Intel Corporation I350 Gigabit Network Connection (rev 01)
04:00.1 Ethernet controller: Intel Corporation I350 Gigabit Network Connection (rev 01)
04:00.2 Ethernet controller: Intel Corporation I350 Gigabit Network Connection (rev 01)
04:00.3 Ethernet controller: Intel Corporation I350 Gigabit Network Connection (rev 01)
09:00.0 Ethernet controller: Intel Corporation Ethernet Controller XL710 for 40GbE QSFP+ (rev 01)
09:00.1 Ethernet controller: Intel Corporation Ethernet Controller XL710 for 40GbE QSFP+ (rev 01)
81:00.0 Ethernet controller: Intel Corporation 82599ES 10-Gigabit SFI/SFP+ Network Connection (rev 01)
81:00.1 Ethernet controller: Intel Corporation 82599ES 10-Gigabit SFI/SFP+ Network Connection (rev 01)
86:00.0 Ethernet controller: Intel Corporation Device 15a4
87:00.0 Ethernet controller: Intel Corporation Device 15a4

[root@RRC1 sh]# lspci -v -s 87:00.0
87:00.0 Ethernet controller: Intel Corporation Device 15a4
        Subsystem: Intel Corporation Device 0000
        Physical Slot: 9
        Flags: bus master, fast devsel, latency 0
        Memory at 3800e4000000 (64-bit, prefetchable) [size=4M]
        Memory at 3800ec400000 (64-bit, prefetchable) [size=8K]
        Memory at 3800e0000000 (64-bit, prefetchable) [size=64M]
        Capabilities: [40] Power Management version 3
        Capabilities: [70] Express Endpoint, MSI 00
        Capabilities: [b0] MSI-X: Enable+ Count=256 Masked-
        Capabilities: [d0] Vital Product Data
        Capabilities: [100] Advanced Error Reporting
        Capabilities: [148] Device Serial Number a0-36-9f-ff-ff-60-b7-2c
        Capabilities: [158] Alternative Routing-ID Interpretation (ARI)
        Capabilities: [168] #19
        Capabilities: [188] Single Root I/O Virtualization (SR-IOV)
        Kernel driver in use: igb_uio
        Kernel modules: fm10k

I assign the pci 86:00.0 to the VM. Here is the output info in the VM:
[root@localhost sh]# lspci | grep Eth
00:03.0 Ethernet controller: Red Hat, Inc Virtio network device
00:08.0 Ethernet controller: Intel Corporation Device 15a4
# [root@localhost tools]# ./cpu_layout.py 
# Core and Socket Information (as reported by '/proc/cpuinfo')

cores =  [0, 1, 2, 3]
sockets =  [0]

```
   Socket 0      
   --------      
```

Core 0 [0, 1]        

Core 1 [2, 3]        

Core 2 [4, 5]        

Core 3 [6, 7]        

run the setup.sh script file to init:
[root@localhost pktgen-2.9.5]# ./setup.sh 
Using directory: /root/sh/dpdk-2.1.0/x86_64-native-linuxapp-gcc
Setup 1 socket(s) with 2048 pages.
AnonHugePages:         0 kB
HugePages_Total:    2048
HugePages_Free:     2048
HugePages_Rsvd:        0
HugePages_Surp:        0
Hugepagesize:       2048 kB
trying to remove old igb_uio module and may get an error message, ignore it
rmmod: ERROR: Module igb_uio is not currently loaded
trying to remove old rte_kni module and may get an error message, ignore it
rmmod: ERROR: Module rte_kni is not currently loaded
# Network devices using DPDK-compatible driver

0000:00:08.0 'Device 15a4' drv=igb_uio unused=fm10k
# Network devices using kernel driver

0000:00:03.0 'Virtio network device' if= drv=virtio-pci unused=virtio_pci,igb_uio 
# Other network devices

<none>

Command 'lspci | grep Ether' output for reference
00:03.0 Ethernet controller: Red Hat, Inc Virtio network device
00:08.0 Ethernet controller: Intel Corporation Device 15a4

Here is the test/set_seq.pkt file:
# seq <seq#> <portlist> dst-Mac src-Mac dst-IP src-IP/netmask sport dport ipv4|ipv6|vlan udp|tcp|icmp vid pktsize

seq 0 all 0001:7000:0001 a036:9f60:b72a 70.0.0.1 70.0.0.22/24 1234 5678 ipv4 udp 1 128
set all seqCnt 1

Then run the pktgen command:
./app/app/${target}/pktgen -c 0xf -n 2 --proc-type auto --socket-mem 2048 -b 00:03.0 -- -p 0x1 -T -P -m "[1:3].0" -f test/set_seq.pkt

The output info is:

EAL: Master lcore 0 is ready (tid=3af458c0;cpuset=[0])
EAL: lcore 3 is ready (tid=b71fd700;cpuset=[3])
EAL: lcore 2 is ready (tid=b79fe700;cpuset=[2])
EAL: lcore 1 is ready (tid=b81ff700;cpuset=[1])
EAL: PCI device 0000:00:03.0 on NUMA socket -1
EAL:   probe driver: 1af4:1000 rte_virtio_pmd
EAL:   Device is blacklisted, not initializing
EAL: PCI device 0000:00:08.0 on NUMA socket -1
EAL:   probe driver: 8086:15a4 rte_pmd_fm10k
EAL:   PCI memory mapped at 0x7fc038800000
EAL:   PCI memory mapped at 0x7fc038c00000
EAL:   PCI memory mapped at 0x7fbfa4000000
[1:3].0          lcores: RX( 1 )TX( 3 ) ports: RX( 0 )TX( 0 )

Lua 5.3.0  Copyright (C) 1994-2015 Lua.org, PUC-Rio

> > > Packet Burst 32, RX Desc 512, TX Desc 512, mbufs/port 4096, mbuf cache 512

=== port to lcore mapping table (# lcores 4) ===
   lcore:     0     1     2     3 
port   0:  D: T  1: 0  0: 0  0: 1 =  1: 1
Total   :  0: 0  1: 0  0: 0  0: 1
    Display and Timer on lcore 0, rx:tx counts per port/lcore

Configuring 1 ports, MBUF Size 1920, MBUF Cache Size 512
Lcore:
    1, RX-Only
                RX( 1): ( 0: 0) 
    3, TX-Only
                TX( 1): ( 0: 0) 

Port :
    0, nb_lcores  2, private 0x8ad230, lcores:  1  3 

*\* Dev Info (rte_pmd_fm10k:0) **
   max_vfs        :   0 min_rx_bufsize    : 256 max_rx_pktlen : 15360 max_rx_queues         : 128 max_tx_queues: 128
   max_mac_addrs  :  64 max_hash_mac_addrs:   0 max_vmdq_pools:    64
   rx_offload_capa:  15 tx_offload_capa   :  15 reta_size     :   128 flow_type_rss_offloads:0000000000000000
   vmdq_queue_base:   0 vmdq_queue_num    :   0 vmdq_pool_base:     0
*\* RX Conf **
   pthreash       :   8 hthresh          :   8 wthresh        :     0
   Free Thresh    :  32 Drop Enable      :   0 Deferred Start :     0
*\* TX Conf **
   pthreash       :  32 hthresh          :   0 wthresh        :     0
   Free Thresh    :  32 RS Thresh        :  32 Deferred Start :     0 TXQ Flags:00000f01

PMD: fm10k_dev_configure(): fm10k always strip CRC
Initialize Port 0 -- TxQ 1, RxQ 1,  Src MAC a0:36:9f:60:b7:2a
    Create: Default RX  0:0  - Memory used (MBUFs 4096 x (size 1920 + Hdr 128)) + 1581248 =   9737 KB

```
Create: Default TX  0:0  - Memory used (MBUFs 4096 x (size 1920 + Hdr 128)) + 1581248 =   9737 KB
Create: Range TX    0:0  - Memory used (MBUFs 4096 x (size 1920 + Hdr 128)) + 1581248 =   9737 KB
Create: Sequence TX 0:0  - Memory used (MBUFs 4096 x (size 1920 + Hdr 128)) + 1581248 =   9737 KB
Create: Special TX  0:0  - Memory used (MBUFs   64 x (size 1920 + Hdr 128)) + 1581248 =   1673 KB

                                                                   Port memory used =  40617 KB
                                                                  Total memory used =  40617 KB
```

Port  0: Link Up - speed 0 Mbps - full-duplex <Enable promiscuous mode>

=== Display processing on lcore 0
WARNING: Nothing to do on lcore 2: exiting
=== RX processing on lcore  1, rxcnt 1, port/qid, 0/0
=== TX processing on lcore  3, txcnt 1, port/qid, 0/0

Please notice the output:"Port  0: Link Up - speed 0 Mbps - full-duplex <Enable promiscuous mode>", the app can't get the correct link speed.

Then use the command to start the port to send the packets:
Pktgen>start all
!ERROR!: Unknown EtherType 0x0000
!ERROR!: Unknown EtherType 0x0000
!ERROR!: Unknown EtherType 0x0000
!ERROR!: Unknown EtherType 0x0000
!ERROR!: Unknown EtherType 0x0000
!ERROR!: Unknown EtherType 0x0000
!ERROR!: Unknown EtherType 0x0000
!ERROR!: Unknown EtherType 0x0000
!ERROR!: Unknown EtherType 0x0000
!ERROR!: Unknown EtherType 0x0000
!ERROR!: Unknown EtherType 0x0000
!ERROR!: Unknown EtherType 0x0000
!ERROR!: Unknown EtherType 0x0000
./doit.sh: line 54:  2402 Segmentation fault      (core dumped) ./app/app/${target}/pktgen -c 0xf -n 2 --proc-type auto --socket-mem 2048 -b 00:03.0 -- -p 0x1 -T -P -m "[1:3].0" -f test/set_seq.pkt

Is there something i forget to configure the port?  Can you give me some help?  many thanks!

**Comment by pktgen** [comment]
Hi

You do not need to use -p 0x1 in the command as it is ignored now. 

Is that nic supported by DPDK? I am not able to check the docs to make sure. But not getting the speed is a problem and I do not have any of those cards. 

Sorry everything looked ok

**Comment by hufangyin** [comment]
OK, I will clear the -p 0x1 , the DPDK version is 2.1.0, and this version is support the FM10K .

**Comment by hufangyin** [comment]
the pktgen have a command:

capture <portlist> <state>       this command can capture the packets? but where is the capture packets file saved???

**Comment by pktgen** [comment]
When you turn off the capture the file is saved in the same place where pktgen was started from

**Comment by hufangyin** [comment]
I don't see any capture file saved in my pc. now the pktgen is start to send the packets, but the port 0 link speed is also 0 and the pktgen send 32 packets interval 5 seconds, how can i set the pktgen send packet continues?

Ports 0-0 of 1   <Main Page>  Copyright (c) <2010-2015>, Wind River Systems
  Flags:Port    :   P-------------:0
Link State      :          <UP-0-FD>    ---TotalRate--- 
Pkts/s  Rx      :                  0                  0 
        Tx      :                 32                 32 
MBits/s Rx/Tx   :                0/0                0/0 
Broadcast       :                  0
Multicast       :                  0
  64 Bytes      :                  0
  65-127        :                  0
  128-255       :                  0
  256-511       :                  0
  512-1023      :                  0
  1024-1518     :                  0
Runts/Jumbos    :                0/0
Errors Rx/Tx    :                0/0
Total Rx Pkts   :                  0
      Tx Pkts   :                896
      Rx MBs    :                  0
      Tx MBs    :                  0
ARP/ICMP Pkts   :                0/0
                :
Tx Count/% Rate :     Forever / 100%
PktSize/Tx Burst:          64 /   32
Src/Dest Port   :        1234 / 5678
Pkt Type:VLAN ID:    IPv4 / TCP:0001
Dst  IP Address :        192.168.1.1
Src  IP Address :     192.168.0.1/24
Dst MAC Address :  00:00:00:00:00:00
Src MAC Address :  a0:36:9f:60:b7:2a
-- Pktgen Ver:2.9.5(DPDK-2.1.0)  Powered by IntelÂ® DPDK ----------------------

**Comment by hufangyin** [comment]
In the Pktgen cmdline, what command can modify the Dst MAC addr?  

/ Ports 0-0 of 1   <Main Page>  Copyright (c) <2010-2015>, Wind River Systems
  Flags:Port    :   P-------------:0
Link State      :      <UP-40000-FD>    ---TotalRate--- 
Pkts/s  Rx      :                  0                  0 
        Tx      :                  0                  0 
MBits/s Rx/Tx   :                0/0                0/0 
Broadcast       :                  0
Multicast       :                  0
  64 Bytes      :                  0
  65-127        :                  0
  128-255       :                  0
  256-511       :                  0
  512-1023      :                  0
  1024-1518     :                  0
Runts/Jumbos    :                0/0
Errors Rx/Tx    :                0/0
Total Rx Pkts   :                  0
      Tx Pkts   :                  0
      Rx MBs    :                  0
      Tx MBs    :                  0
ARP/ICMP Pkts   :                0/0
                :
Tx Count/% Rate :          10 /   1%
PktSize/Tx Burst:          64 /   32
Src/Dest Port   :        1234 / 5678
Pkt Type:VLAN ID:    IPv4 / TCP:0001
Dst  IP Address :           70.0.0.1
Src  IP Address :       70.0.0.22/24
Dst MAC Address :  00:00:00:00:00:00
Src MAC Address :  a0:36:9f:60:b7:2a
-- Pktgen Ver:2.9.5(DPDK-2.1.0)  Powered by IntelÂ® DPDK ----------------------
 
 
 
 
 
 
 
 
Pktgen > 
Pktgen> 
Pktgen> set 0 rate 1
Pktgen> set 0 count 10
Pktgen> set ip dst 0 70.0.0.1
Pktgen> set ip src 0 70.0.0.22/24

I use the command modify the fields but the Dst MAC Address is also 00:00:00:00:00:00, how can i use the command to modify the Dst MAC Address as 00:11:22:33:44:55 ????

**Comment by pktgen** [comment]
Pktgen only captures rx packets not tx packets. Did you enable capture, rx some packets then turn off capture?

**Comment by pktgen** [comment]
The command is set mac 0 1234:1234:1234

0 is the port list look in the help screens for more info

**Comment by hufangyin** [comment]
hi, i use the pktgen command set the size of the packet to transmit, but the size of the packet to send is not change.

Pktgen> set 0 size 128

but the size of packet is also 64 bytes.

**Comment by pktgen** [comment]
The command works for the single packet mode which is the size displayed on the main page. If you are trying to use sequence or range packets then you need to use the commands for that function. Look in the help section for the other commands and let me know if that was the problem.

**Comment by przemeklal** [comment]
Hi,
did you manage to resolve issues with pktgen on this setup described above? I'm trying to run pktgen on VM with fm10k virtual functions and I've got into exactly the same problems.

**Comment by twistUser1** [comment]
I compiled pktgen 2.9.11 version over DPDK 16.04.
I noticed that it only sends 32 packet per second. 
I configured "set 0,1 rate 10" to send traffic at 1G line rate. But it is still sending at 32 frames per second.
Can we use it to send continous traffic ? Can anyone share any configuration required for this?

Can we use it to send traffic at 10G line rate for packet size 256 for ex,, i.e rate of 844594 frames per second.
